core/src/main/java/org/bouncycastle/crypto/digests/Blake3Digest.java

```diff
@@ -0,0 +1,907 @@
+
+        /* Loop through the rounds */
+        for (int round = 0; round < ROUNDS - 1; round++)
+        {
+            /* Perform the round and permuteM */
+            performRound();
+            permuteIndices();
+        }
+        performRound();
+        adjustChaining();
+    }
+
+    /**
+     * Perform a round.
+     */
+    private void performRound()
+    {
+        /* Apply to columns of V */
+        int idx = 0;
+        mixG(idx++, CHAINING0, CHAINING4, IV0, COUNT0);
+        mixG(idx++, CHAINING1, CHAINING5, IV1, COUNT1);
+        mixG(idx++, CHAINING2, CHAINING6, IV2, DATALEN);
+        mixG(idx++, CHAINING3, CHAINING7, IV3, FLAGS);
+
+        /* Apply to diagonals of V */
+        mixG(idx++, CHAINING0, CHAINING5, IV2, FLAGS);
+        mixG(idx++, CHAINING1, CHAINING6, IV3, COUNT0);
+        mixG(idx++, CHAINING2, CHAINING7, IV0, COUNT1);
+        mixG(idx, CHAINING3, CHAINING4, IV1, DATALEN);
+    }
+
+    /**
+     * Initialise M from message.
+     * @param pMessage the source message
+     * @param pMsgPos the message position
+     */
+    private void initM(final byte[] pMessage,
+                       final int pMsgPos)
+    {
+        /* Copy message bytes into word array */
+        for (int i = 0; i < NUMWORDS << 1; i++)
+        {
+            theM[i] = Pack.littleEndianToInt(pMessage, pMsgPos + i * Integer.BYTES);
+        }
+    }
+
+    /**
+     * Adjust Chaining after compression.
+     */
+    private void adjustChaining()
+    {
+        /* If we are outputting */
+        if (outputting)
+        {
+            /* Adjust full state */
+            for (int i = 0; i < NUMWORDS; i++)
+            {
+                theV[i] ^= theV[i + NUMWORDS];
+                theV[i + NUMWORDS] ^= theChaining[i];
+            }
+
+            /* Output state to buffer */
+            for (int i = 0; i < NUMWORDS << 1; i++)
+            {
+                Pack.intToLittleEndian(theV[i], theBuffer, i * Integer.BYTES);
+            }
+            thePos = 0;
+
+            /* Else just build chain value */
+        }
+        else
+        {
+            /* Combine V into Chaining */
+            for (int i = 0; i < NUMWORDS; i++)
+            {
+                theChaining[i] = theV[i] ^ theV[i + NUMWORDS];
+            }
+        }
+    }
+
+    /**
+     * Mix function G.
+     * @param msgIdx the message index
+     * @param posA position A in V
+     * @param posB position B in V
+     * @param posC position C in V
+     * @param posD poistion D in V
+     */
+    private void mixG(final int msgIdx,
+                      final int posA,
+                      final int posB,
+                      final int posC,
+                      final int posD)
+    {
+        /* Determine indices */
+        int msg = msgIdx << 1;
+        int rot = 0;
+
+        /* Perform the Round */
+        theV[posA] += theV[posB] + theM[theIndices[msg++]];
+        theV[posD] = rotr32(theV[posD] ^ theV[posA], ROTATE[rot++]);
```

Comment:
Nit: this can use `Integer.rotateRight`, available since Java 1.5.

Reply:
Thanks for the comment. At the moment it's better like this, an even bigger nit we have to deal with is we have people using Java 4... Sun licensed so many JVMs, it's quite a broad church still.